malsim/sims/mal_simulator.py

```diff
@@ -566,126 +649,124 @@ def _defender_step(self, agent, defense_step):
-        infos = {}
+        # TODO: need to handle defense steps disabling attack steps
+        if not self.sim_settings.cumulative_defender_obs:
+            # View all states as unknown
+            # TODO: Should they be disabled (0) instead?
+            obs_state.fill(-1)
 
-        can_wait = {
-            "attacker": 0,
-            "defender": 1,
-        }
+        # Only show the latest steps taken
+        for _, actions in performed_actions.items():
+            for action in actions:
+                obs_state[action] = 1
+
+    def _observe_agents(self, performed_actions):
+        """Collect agents observations"""
 
-        finished_agents = []
-        # If no attackers have any actions left that they could take the
-        # simulation will terminate.
-        attackers_done = True
-        # Fill in the agent observations, rewards, terminations, truncations,
-        # and infos.
         for agent in self.agents:
-            # Collect agent observations
-            agent_observation = copy.deepcopy(self._blank_observation)
-            if self.agents_dict[agent]["type"] == "defender":
-                self._observe_defender(
-                    agent, performed_actions, agent_observation)
-            elif self.agents_dict[agent]["type"] == "attacker":
-                self._observe_attacker(agent, agent_observation)
+            agent_type = self.agents_dict[agent]["type"]
+            if  agent_type == "defender":
+                self._observe_defender(agent, performed_actions)
+
+            elif agent_type == "attacker":
+                self._observe_attacker(agent, performed_actions)
+
             else:
                 logger.error(
                     "Agent %s has unknown type: %s",
                     agent, self.agents_dict[agent]["type"]
                 )
 
-            observations[agent] = agent_observation
+    def _reward_agents(self, performed_actions):
+        """Update rewards from latest performed actions"""
+        # TODO: should initial actions give rewards? Currently do.
+        for agent, actions in performed_actions.items():
+            agent_type = self.agents_dict[agent]["type"]
+
+            for action in actions:
+                if action is None:
+                    continue
+
+                node_id = self._index_to_id[action]
+                node = self.attack_graph.get_node_by_id(node_id)
+                node_reward = node.extras.get('reward', 0)
+
+                if agent_type == "attacker":
+                    # If attacker performed step, it will receive
+                    # a reward and penalize all defender
+                    self._agent_rewards[agent] += node_reward
+
+                    for d_agent in self.get_defender_agents():
+                        self._agent_rewards[d_agent] -= node_reward
+                else:
+                    # TODO: defending against an attack step
+                    # can it increase reward for defender?
+
```

Comment:
Well, technically it increases the penalty, since it's negative. Apart from that, yes, that's how defences work. They cost something, so the defender is penalized if there is a reward associated with it. Which makes sense, if there's any defence without a cost then the correct action is just to always enable it immediately.